clang/lib/Sema/CheckExprLifetime.cpp

```diff
@@ -249,9 +254,10 @@ static void visitLocalsRetainedByReferenceBinding(IndirectLocalPath &Path,
                                                   LocalVisitor Visit);
 
 template <typename T> static bool isRecordWithAttr(QualType Type) {
-  if (auto *RD = Type->getAsCXXRecordDecl())
-    return RD->hasAttr<T>();
-  return false;
+  CXXRecordDecl *RD = Type.getNonReferenceType()->getAsCXXRecordDecl();
```

Comment:
I am a bit worried here. While `lifetime_capture_by` applies to the value behind the reference, `lifetimebound` always apply to the top level construct (so the reference itself). So I am wondering if this change is correct when we process a lifetimebound annotation.

Reply:
I do not think there is a difference in the annotations in that respect.
For example, consider:
```cpp
std::string_view foo(std::string_view&& t [[clang::lifetimebound]]);

void use() {
    std::string_view x = foo(std::string_view()); // Clangs warns here.
}
```
I think this is, in principle, a false positive. Pointer-like types when appearing as a reference type should be considered values.
https://github.com/llvm/llvm-project/issues/116066